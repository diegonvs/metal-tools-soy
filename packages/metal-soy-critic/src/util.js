const chalk = require('chalk');
const Promise = require('bluebird');

function combineResults(first, second) {
  return toResult(
    first.status && second.status,
    ...first.messages.concat(second.messages)
  );
}

function difference(setA, setB) {
  const difference = new Set();
  for (const elem of setA) {
    if (!setB.has(elem)) {
      difference.add(elem);
    }
  }
  return difference;
}

function toResult(status, ...messages) {
  return {
    messages,
    status
  };
}

function parseTemplateName(rawName) {
  const segments = rawName.split('.');
  const namespace = segments
    .slice(0, segments.length - 1)
    .join('.');

  return {
    name: segments[segments.length - 1],
    namespace: namespace || null
  };
}

function sequence(...tasks) {
  return Promise.reduce(tasks, (res, next) =>
    next().then(value => [...res, value]),
    []
  );
}

function joinErrors(lines) {
  return lines
    .map(line => chalk.red(line))
    .join('\n');
}

module.exports = {
  combineResults,
  difference,
  joinErrors,
  parseTemplateName,
  sequence,
  toResult
};
