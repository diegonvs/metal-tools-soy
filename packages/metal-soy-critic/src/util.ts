import * as chalk from 'chalk';

export interface Result {
  status: boolean;
  messages: Array<string>;
}

export function combineResults(first: Result, second: Result): Result {
  return toResult(
    first.status && second.status,
    ...first.messages.concat(second.messages)
  );
}

export function toResult(status: boolean, ...messages: Array<string>): Result {
  return {
    messages,
    status
  };
}

export interface TemplateName {
  name: string,
  namespace: string | null
}

export function difference<T>(setA: Set<T>, setB: Set<T>): Set<T> {
  const difference = new Set();
  for (const elem of setA) {
    if (!setB.has(elem)) {
      difference.add(elem);
    }
  }
  return difference;
}

export function joinErrors(lines: Array<string>): string {
  return lines
    .map(line => chalk.red(line))
    .join('\n');
}

export function isSorted(items: Array<string>): boolean {
  const sortedItems = items
    .concat()
    .sort((a, b) => a.localeCompare(b));

  return sortedItems.join() === items.join();
}
