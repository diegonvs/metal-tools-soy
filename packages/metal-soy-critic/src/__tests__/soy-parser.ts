import {getFixture} from '../../test/util';
import soyParser from '../soy-parser';

describe('Soy Parser', () => {
  test('should parse file correctly', () => {
    const input = getFixture('Test.soy');

    expect(soyParser(input)).toMatchSnapshot();
  });

  test('should fail to parse file', () => {
    const input = getFixture('Invalid.soy');

    expect(() => soyParser(input)).toThrow();
  });

  test('should parse nested calls', () => {
    const input = getFixture('NestedCalls.soy');

    expect(soyParser(input)).toMatchSnapshot();
  });

  test('should parse deltemplates', () => {
    const input = getFixture('DelTemplate.soy');

    expect(soyParser(input)).toMatchSnapshot();
  });

  test('should parse private templates', () => {
    const input = getFixture('PrivateTemplate.soy');

    expect(soyParser(input)).toMatchSnapshot();
  });
});
