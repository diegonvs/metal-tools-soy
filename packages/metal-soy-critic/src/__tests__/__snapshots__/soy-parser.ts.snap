// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Soy Parser should parse deltemplates 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "content": "$title",
          "type": "Interpolation",
        },
        Object {
          "content": "$name",
          "type": "Interpolation",
        },
        Object {
          "body": Array [
            Object {
              "content": "$optionalInfo",
              "type": "Interpolation",
            },
          ],
          "type": "If",
        },
        Object {
          "body": Array [],
          "name": "helper",
          "namespace": null,
          "type": "Call",
        },
        Object {
          "body": Array [
            Object {
              "body": Array [],
              "name": "foo",
              "type": "Param",
            },
          ],
          "name": "render",
          "namespace": "OtherComponent",
          "type": "Call",
        },
      ],
      "name": "First",
      "namespace": "Test",
      "params": Array [
        Object {
          "name": "name",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "title",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "optionalInfo",
          "paramType": "string",
          "required": false,
          "type": "ParamDeclaration",
        },
      ],
      "type": "DelTemplate",
      "variant": null,
    },
    Object {
      "body": Array [
        Object {
          "content": "$title",
          "type": "Interpolation",
        },
        Object {
          "content": "$name",
          "type": "Interpolation",
        },
        Object {
          "body": Array [
            Object {
              "content": "$optionalInfo",
              "type": "Interpolation",
            },
          ],
          "type": "If",
        },
        Object {
          "body": Array [],
          "name": "helper",
          "namespace": null,
          "type": "Call",
        },
        Object {
          "body": Array [
            Object {
              "body": Array [],
              "name": "foo",
              "type": "Param",
            },
          ],
          "name": "render",
          "namespace": "OtherComponent",
          "type": "Call",
        },
      ],
      "name": "Second",
      "namespace": "Othernamespace",
      "params": Array [
        Object {
          "name": "name",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "title",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "optionalInfo",
          "paramType": "string",
          "required": false,
          "type": "ParamDeclaration",
        },
      ],
      "type": "DelTemplate",
      "variant": Object {
        "content": "'neato'",
        "type": "Interpolation",
      },
    },
  ],
  "namespace": "Test",
  "type": "Program",
}
`;

exports[`Soy Parser should parse file correctly 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "content": "$title",
          "type": "Interpolation",
        },
        Object {
          "content": "$name",
          "type": "Interpolation",
        },
        Object {
          "body": Array [
            Object {
              "content": "$optionalInfo",
              "type": "Interpolation",
            },
          ],
          "type": "If",
        },
        Object {
          "body": Array [],
          "name": "helper",
          "namespace": null,
          "type": "Call",
        },
        Object {
          "body": Array [
            Object {
              "body": Array [],
              "name": "foo",
              "type": "Param",
            },
          ],
          "name": "render",
          "namespace": "OtherComponent",
          "type": "Call",
        },
      ],
      "name": "render",
      "namespace": null,
      "params": Array [
        Object {
          "name": "name",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "title",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "optionalInfo",
          "paramType": "string",
          "required": false,
          "type": "ParamDeclaration",
        },
      ],
      "private": false,
      "type": "Template",
    },
    Object {
      "body": Array [],
      "name": "helper",
      "namespace": null,
      "params": Array [
        Object {
          "name": "test",
          "paramType": "asdasd",
          "required": false,
          "type": "ParamDeclaration",
        },
      ],
      "private": false,
      "type": "Template",
    },
  ],
  "namespace": "Test",
  "type": "Program",
}
`;

exports[`Soy Parser should parse nested calls 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "content": "$title",
          "type": "Interpolation",
        },
        Object {
          "content": "$name",
          "type": "Interpolation",
        },
        Object {
          "body": Array [
            Object {
              "content": "$optionalInfo",
              "type": "Interpolation",
            },
          ],
          "type": "If",
        },
        Object {
          "body": Array [
            Object {
              "content": "$name",
              "type": "Interpolation",
            },
            Object {
              "content": "$title",
              "type": "Interpolation",
            },
          ],
          "type": "If",
        },
        Object {
          "body": Array [
            Object {
              "body": Array [],
              "name": "foo",
              "type": "Param",
            },
            Object {
              "body": Array [
                Object {
                  "body": Array [],
                  "name": "helper",
                  "namespace": null,
                  "type": "Call",
                },
                Object {
                  "body": Array [],
                  "name": "helper",
                  "namespace": null,
                  "type": "Call",
                },
              ],
              "name": "content",
              "type": "Param",
            },
          ],
          "name": "render",
          "namespace": "OtherComponent",
          "type": "Call",
        },
        Object {
          "body": Array [],
          "name": "helper",
          "namespace": null,
          "type": "Call",
        },
      ],
      "name": "render",
      "namespace": null,
      "params": Array [
        Object {
          "name": "name",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "title",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "optionalInfo",
          "paramType": "string",
          "required": false,
          "type": "ParamDeclaration",
        },
      ],
      "private": false,
      "type": "Template",
    },
    Object {
      "body": Array [],
      "name": "helper",
      "namespace": null,
      "params": Array [
        Object {
          "name": "test",
          "paramType": "asdasd",
          "required": false,
          "type": "ParamDeclaration",
        },
      ],
      "private": false,
      "type": "Template",
    },
  ],
  "namespace": "Test",
  "type": "Program",
}
`;

exports[`Soy Parser should parse param declarations 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "body": Array [
            Object {
              "content": "$title",
              "type": "Interpolation",
            },
          ],
          "type": "Foreach",
        },
        Object {
          "content": "$name",
          "type": "Interpolation",
        },
        Object {
          "body": Array [
            Object {
              "content": "$optionalInfo",
              "type": "Interpolation",
            },
          ],
          "type": "If",
        },
      ],
      "name": "render",
      "namespace": null,
      "params": Array [
        Object {
          "name": "name",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "titles",
          "paramType": "list<string>",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "optionalInfo",
          "paramType": "html|string",
          "required": false,
          "type": "ParamDeclaration",
        },
      ],
      "private": false,
      "type": "Template",
    },
  ],
  "namespace": "Test",
  "type": "Program",
}
`;

exports[`Soy Parser should parse param names with underscores 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "content": "$_name",
          "type": "Interpolation",
        },
      ],
      "name": "render",
      "namespace": null,
      "params": Array [
        Object {
          "name": "_name",
          "paramType": "string",
          "required": false,
          "type": "ParamDeclaration",
        },
      ],
      "private": false,
      "type": "Template",
    },
  ],
  "namespace": "UnderscoreParam",
  "type": "Program",
}
`;

exports[`Soy Parser should parse private templates 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "content": "$title",
          "type": "Interpolation",
        },
        Object {
          "content": "$name",
          "type": "Interpolation",
        },
        Object {
          "body": Array [],
          "name": "helper",
          "namespace": null,
          "type": "Call",
        },
      ],
      "name": "render",
      "namespace": null,
      "params": Array [
        Object {
          "name": "name",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
        Object {
          "name": "title",
          "paramType": "string",
          "required": true,
          "type": "ParamDeclaration",
        },
      ],
      "private": false,
      "type": "Template",
    },
    Object {
      "body": Array [
        Object {
          "content": "$test ?: ''",
          "type": "Interpolation",
        },
      ],
      "name": "helper",
      "namespace": null,
      "params": Array [
        Object {
          "name": "test",
          "paramType": "asdasd",
          "required": false,
          "type": "ParamDeclaration",
        },
      ],
      "private": true,
      "type": "Template",
    },
  ],
  "namespace": "Test",
  "type": "Program",
}
`;
